### PR TITLE
feat: <1 year GitHub red flag for new joiner

### PR DESCRIPTION
**Pull request by ravikumar1002**
### What is the change?
 Added red flag if  user GitHub account created within 1 year and if user Github account is more then 1 year then it'll show 
 in grey color
 
   - Make a utility function that will give you the difference between a start date and an end date. The return value will be an object with the key of years and months.
   - showed message according to  GitHub created an account.

### Backend changes
https://github.com/Real-Dev-Squad/website-backend/pull/1362

### Frontend UI test

 issue: It's hard, so created new ticket https://github.com/Real-Dev-Squad/website-www/issues/558

### Is Development Tested?
- Yes
 
image
image

### Before :
image

### After :

image
image
image

https://github.com/Real-Dev-Squad/website-www/assets/86847625/0c9d8c6e-3cac-434d-ade3-fea03a96be83

### CODE REVIEW

File: js/intro.js [hunk 1]
```diff
@@ -240,6 +256,15 @@ async function showSavedDetails() {
   try {
     const userId = urlParams.get('id');
     const usersRequest = await makeApiCall(`${BASE_URL}/users/${userId}/intro`);
+    const userInformation = await makeApiCall(
+      `${BASE_URL}/users/userId/${userId}`,
+    );
+
+    const dateDiff = getDateDifferenceInYearsAndMonths(
+      userInformation?.data.user.github_created_at,
```
Comment (sahsisunny) [inline js/intro.js:264]: if `github_created_at` is not found then?

**Comment by sahsisunny** [comment]
This PR depends on another backend PR, right? Can you mention it in the main description?

**Review by vinit717 — CHANGES_REQUESTED** [review]
Tests?

File: js/intro.js (hunk 1 above)
Comment (ravikumar1002) [reply js/intro.js:264]: This will return NaN but will not break UI as I have used optional chaining and have conditional rendering in frontend part

**Comment by ravikumar1002** [comment]
> This PR depends on another backend PR, right? Can you mention it in the main description?

Ok

**Comment by ravikumar1002** [comment]
> Tests?
There is no setup for testing currently.

**Comment by vinit717** [comment]
> > Tests?
> 
> There is no setup for testing currently.

without test, no PR is getting merged if you can't ask @ankushdharkar for this

File: js/intro.js (hunk 1 above)
Comment (ravikumar1002) [reply js/intro.js:264]: OK, now understand, I will fix this

**Comment by ravikumar1002** [comment]
> > > Tests?
> > 
> > 
> > There is no setup for testing currently.
> 
> without test, no PR is getting merged if you can't ask @ankushdharkar for this

but there is no test for this repo, I asked sunny for this previously, and every PR getting merged without testing for this with the development test

File: js/intro.js (hunk 1 above)
Comment (ravikumar1002) [reply js/intro.js]: done

File: __test__/utils.test.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+const utils = require('../js/utils');
+
+describe('Utils functiona test case', () => {
```
Comment (kotesh-arya) [inline __test__/utils.test.js:3]: `functiona` => `function` 👍🏻

File: js/intro.js [hunk 3]
```diff
@@ -64,7 +64,7 @@ function generateNoDataFoundPage() {
   notFound.appendChild(notFoundDiv);
 }
 
-function generateSavedDetailsForm(users) {
+function generateSavedDetailsForm(users, dateDiff = NaN) {
```
Comment (kotesh-arya) [inline js/intro.js:67]: `generateSavedDetailsForm` => `generateSavedDetailsFrom`
As we are generating saved Details **FROM** the passed arguments.

File: js/intro.js [hunk 4]
```diff
@@ -257,7 +289,7 @@ async function showSavedDetails() {
         numberOfHours: userData.intro.numberOfHours,
         foundFrom: userData.foundFrom,
       };
-      generateSavedDetailsForm(userSavedData);
+      generateSavedDetailsForm(userSavedData, dateDiff);
```
Comment (kotesh-arya) [inline js/intro.js:292]: Here too, the function name should be Obvoius 👍🏻

File: __test__/index.test.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-const timeout = 15000;
+const timeout = 25000;
 
 describe('Dummy Test ', () => {
   beforeAll(async () => {
```
Comment (vinit717) [inline __test__/index.test.js:4]: why did you change the timeout  ? you are not written any test here

File: js/intro.js [hunk 6]
```diff
@@ -77,6 +77,29 @@ function generateSavedDetailsForm(users) {
...
+    dateDiff.months === 0 &&
+    dateDiff.days >= 0
+  ) {
+    githubCreatedValue.innerText = `User GitHub account created ${
+      dateDiff.days > 0 ? dateDiff.days : '1'
+    } days ago`;
+  }
+
+  container.appendChild(githubCreatedValue);
+
```
Comment (vinit717) [inline js/intro.js:102]: Can you test the classes and page UI as puppeteer can test this

File: __test__/index.test.js (hunk 5 above)
Comment (ravikumar1002) [reply __test__/index.test.js:4]: because test failing due to time limit that's why

File: js/intro.js (hunk 6 above)
Comment (ravikumar1002) [reply js/intro.js:102]: This is not a user interface, which means user is not able to click anywhere it's just a text

File: __test__/utils.test.js (hunk 2 above)
Comment (ravikumar1002) [reply __test__/utils.test.js:3]: ok I will correct this

Comment (ravikumar1002) [reply __test__/utils.test.js:3]: ok

File: js/intro.js (hunk 6 above)
Comment (vinit717) [reply js/intro.js:102]: puppeteer can test the text so test the text also we are already testing the test and classes in dashboard site

File: __test__/index.test.js (hunk 5 above)
Comment (vinit717) [reply __test__/index.test.js:4]: why test is failing any specific reason?

File: js/intro.js (hunk 6 above)
Comment (ravikumar1002) [reply js/intro.js]: I have created a new ticket for this

File: __test__/index.test.js (hunk 5 above)
Comment (ravikumar1002) [reply __test__/index.test.js:4]: because of timeout exceeded

File: js/intro.js (hunk 6 above)
Comment (ravikumar1002) [reply js/intro.js]: it's already mentioned

**Review by kotesh-arya — APPROVED** [review]

**Comment by ravikumar1002** [comment]
> Can you provide a video for the flow?

Done

**Review by kotesh-arya — APPROVED** [review]

**Review by vinit717 — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]
